// walker/random/walk_map.js
module.exports = function() {

    if (this.childern.hasOwnProperty('n') && (!this.children.n)) {
        var id = this._id + 'n';
        var points = this.points;
        var last = this.last;
        --last.i;
        points.push(last);
        var value = {
            _id: id,
            points: points,
            last: last,
            children: {
                n: false,
                e: false,
                w: false
            }
        };
        emit(id, value);
    }


    if (this.childern.hasOwnProperty('s') && (!this.children.s)) {
        var id = this._id + 's';
        var points = this.points;
        var last = this.last;
        ++last.i;
        points.push(last);
        var value = {
            _id: id,
            points: points,
            last: last,
            children: {
                s: false,
                e: false,
                w: false
            }
        };
        emit(id, value);
    }


    if (this.childern.hasOwnProperty('e') && (!this.children.e)) {
        var id = this._id + 'e';
        var points = this.points;
        var last = this.last;
        ++last.j;
        points.push(last);
        var value = {
            _id: id,
            points: points,
            last: last,
            children: {
                s: false,
                e: false,
                n: false
            }
        };
        emit(id, value);
    }


    if (this.childern.hasOwnProperty('w') && (!this.children.w)) {
        var id = this._id + 'w';
        var points = this.points;
        var last = this.last;
        --last.j;
        points.push(last);
        var value = {
            _id: id,
            points: points,
            last: last,
            children: {
                s: false,
                w: false,
                n: false
            }
        };
        emit(id, value);
    }

}
