// mola2/ccd.js
var util = require('util');

module.exports = function(image, configs) {
    image.west = parseInt(image.west);
    image.east = parseInt(image.east);
    image.north = parseInt(image.north);
    image.south = parseInt(image.south);

    if (configs) {
        configs.west = parseInt(configs.west);
        configs.east = parseInt(configs.east);
        configs.north = parseInt(configs.north);
        configs.south = parseInt(configs.west);
        var west = Math.max(configs.west, image.west);
        var east = Math.min(configs.east, image.east);
        var north = Math.min(configs.north, image.north);
        var south = Math.max(configs.south, image.south);

        /**
         * keeping in mind that top/bottom goes in a negative direction because north is up
         * and rows go down...
         */
        var left = (image.west - configs.west) * image.scale;
        var top = (image.north - configs.north) * image.scale;
        var right = (image.east - configs.east) * image.scale;
        var bottom = (image.south - configs.south) * image.scale;

        var zoom = parseInt(configs.zoom);
        zoom = Math.max(1, Math.floor(zoom));

        if (zoom > 1) {
            var left_scaled = parseInt(left / zoom);
            var right_scaled = parseInt(right / zoom);
            var bottom_scaled = parseInt(bottom / zoom);
            var top_scaled = parseInt(top / zoom);
        } else {
            var left_scaled = left;
            var right_scaled = right;
            var top_scaled = top;
            var bottom_scaled = bottom;
        }
    }
    else {
        var west = image.west;
        var east = image.east;
        var north = image.north;
        var south = image.south;

        var left = 0;
        var right = 0;
        var top = 0;
        var bottom = 0;

        var left_scaled = 0;
        var right_scaled = 0;
        var top_scaled = 0;
        var bottom_scaled = 0;

        var zoom = 1;
    }

    // the number of rows and columns of input data
    var cols = parseInt((east - west) * image.scale);
    var rows = parseInt((north - south) * image.scale);

    var out = {
        north: north,
        east: east,
        west: west,
        south: south,

        left: left,
        right: right,
        top: top,
        bottom: bottom,

        left_scaled: left_scaled,
        right_scaled: right_scaled,
        top_scaled: top_scaled,
        bottom_scaled: bottom_scaled,

        cols: cols,
        rows: rows,
        zoom: zoom
    }

    console.log(__filename, ': returning ');
    console.log(util.inspect(out));

    return out

}
