// mvc/controller/context/init.js
var bind_view = require('mvc/controller/bind_view');

module.exports = function(Context, gate) {
    Context._defaults = {
        request: null,
        response: null,
        controller: false,
        action: '',
        errors: [],
        next: false,
        handler: function() {},
        view: false
    };

    Context.prototype.init = function(config) {
        for (var prop in config) {
            var item = config[prop];
            switch (prop) {
            case 'request':
            case 'response':
            case 'action':
            case 'next':
            case 'view':
            case 'handler':
            case 'controller':
                this[prop] = item;
                break;
            
            default:
                throw new Error(__filename + ':: contructor - unrecognized property' + prop);
            }
        }

        _.defaults(this, Context._defaults);

        if (this.view && this.action && this.controller) {
            bind_view(this.controller, this.action, this.view);
        }
    }

    gate.task_done();
}
