// mola/index.js
var import_module = require('./import');
var cti         = require('util/canvas/to_image');
var Canvas      = require('canvas');
var color_module = require('util/colors/scale');
var oninc = require('util/oninc');
var fs_utils    = require('util/fs');
var Gate        = require('util/gate');
var fs          = require('fs');
var util        = require('util');
var au          = require('util/array');
var pm          = require('path');
var Math_utils  = require('util/math');

var open_streams = 0;

MOLA = function(path, props) {
    this.path = path;
    this.data = [];
    this.avg = null;
    this.std_dev = null;
    //  this.image_tile_rows = this.get_rows() / 8;
    //  this.image_tile_cols = this.get_cols() / 4;
    for (var p in props) {
        this[p] = props[p];
    }

    this.render_status = true;
    console.log('new MOLA: ');
}

MOLA.extensions = {};

module.exports = MOLA;

MOLA.prototype = {

    export_tiles: require('./export'),

    read: require('./read'),

    tile_data: require('./tile_data'),

    _height_to_color: require('./height_to_color'),

    _height_to_image_data: require('./_height_to_image_data'),

    _tile_col: function(j) {
        return Math.min(this.get_cols(), (this.get_pixels_per_col() * j));
    },

    _tile_row: function(i) {
        return Math.min(this.get_rows(), (this.get_pixels_per_row() * i));
    },

    measure: require('./measure'),
    
    get_scale: function(){
        if (!this.scale) this.scale = require('./scale')();
        return this.scale;
    },

    height_color: function(h) {
        return _.map(color_module.nearest(h, this.get_scale()).rgb, function(c) {
            return parseInt(Math.min(255, Math.max(0, c)));
        }); 
    },

    get_data_count: function() {
        return _.reduce(this.data, function(r, out){ return out + r.length; }, 0);
    },

    as_heightmap: require('./as_heightmap'),
    
    as_canvas: require('./as_canvas'),

    draw: function(png_path, callback) {
        cti(this.as_canvas(), png_path, callback);
    },
    
    as_json: require('./as_json')
    
}

var skip = 0;
