// mola3/grid/Cell.js
var util = require('util');
var _ = require('underscore');

function Cell(parent, row, col, height) {
    this.parent = this.terrain = parent;
    this.row = row;
    this.col = col;
    this.height = height ? height : 0;
    this.neighbors = {};
}

Cell.prototype = {
    each_neighbor:function (do_f) {
        for (var dir in this.neighbors) {
            var neighbor = this.neighbors[dir];
            do_f(this, neighbor, dir);
        }
    },

    get_neighbors: function(){
        var out = [];
        for (var dir in this.neighbors){
            out.push(this.neighbors[dir]);
        }
        return out;
    },

    neighborhood: function(do_f, range){
        var t = this.terrain;
        var self = this;
        t.slice( this.col - range, this.row - range, this.col + range, this.row + range).forEach(function (n){
            do_f(n, self);
        });
    },

    equals: function(n){
        return ((n.row == this.row) && (n.col == this.col));
    }

}

module.exports = Cell;
