// mvc/_meta.js
var fs = require('fs');
var path = require('path');
var meta = require('./../model/meta');

module.exports = exports = {

    index: function(req, res) {
        var cn = [];
        fs.readdir(__dirname,
        function(err, files) {
            if (err) throw err;
            files.forEach(function(file) {
                //       console.log('index: found file ' + file);
                if (/\.js$/.test(file)) {
                    //     console.log('pushing ' + file);
                    cn.push(file.replace(/\.js$/, ''));
                    //   console.log('cn = ');
                    //     console.dir(cn);
                } else {
                    //   console.log('... failed test - skipping');
                }
            });
            var options = {
                locals: {
                    cn: cn,
                    form: require('./../forms/meta/controller').form().render()
                }
            };
            //     console.dir(options);
            res.render(options);
        });
    },

    show: function(req, res, next) {
        var controller_path = __dirname + '/' + req.params.id + '.js';
        var provider_path = __dirname + '/../model/' + req.params.id + '.js';

        var controller = '';
        var provider = '';

        var c_read_path = fs.createReadStream(controller_path);
        c_read_path.on('data',
        function(data) {
            controller += data;
        });

        c_read_path.on('end',
        function() {
            console.log('show end');
            var p_read_path = fs.createReadStream(provider_path);
            p_read_path.on('data',
            function(data) {
                provider += data;
            });
            p_read_path.on('end',
            function() {
                var options = {
                    locals: {
                        provider: provider,
                        controller: controller,
                        name: req.params.id
                    }
                };
                console.log('show - render');
               // console.dir(options);
                res.render(options);
            }); // end p_read_path_on(end);
        }); // c_read_path.on(end);
    }
}

function split_actions(a) {
    var out = [];
    for (var i = 0; i < c.actions.length; ++i) {
        var action = c.actions[i];
        var real_actions = action.split("+");
        for (var r = 0; r < real_actions.length; ++r) {
            out.push(real_actions[r]);
        }
    }
    return out;
}
